File: src/delta_review/gitlab/diffs.py
from __future__ import annotations

from urllib.parse import quote

from delta_review.gitlab.client import GitLabClient, GitLabError
from delta_review.models import DiffFile, MergeRequest


class DiffService:
    def __init__(self, client: GitLabClient) -> None:
        self._client = client

    @staticmethod
    def _merge_request_path(project: str, mr_iid: int) -> str:
        encoded_project = quote(project, safe="")
        return f"/projects/{encoded_project}/merge_requests/{mr_iid}"

    async def get_merge_request(
        self, project: str, mr_iid: int
    ) -> MergeRequest:
        payload = await self._client.request(
            "GET", self._merge_request_path(project, mr_iid)
        )
        return MergeRequest.model_validate(payload)

    async def _raw_changes(
        self, merge_request_path: str
    ) -> list[dict] | None:
        """Read every diff straight from the repository.

        GitLab applies its diff size budget across the whole merge request,
        so files past the budget come back without any content and without a
        ``too_large`` or ``collapsed`` flag. ``access_raw_diffs`` bypasses the
        budget. Returns ``None`` when the retry is unavailable, leaving the
        truncated diffs in place rather than losing them.
        """
        try:
            payload = await self._client.request(
                "GET",
                f"{merge_request_path}/changes",
                params={"access_raw_diffs": "true"},
            )
        except GitLabError:
            return None
        if payload.get("overflow"):
            return None
        return payload.get("changes") or None

    async def get_diffs(self, project: str, mr_iid: int) -> list[DiffFile]:
        merge_request_path = self._merge_request_path(project, mr_iid)
        try:
            raw_files = [
                item
                async for item in self._client.paginate(
                    f"{merge_request_path}/diffs"
                )
            ]
        except GitLabError as error:
            if error.status_code not in {404, 500}:
                raise
            payload = await self._client.request(
                "GET", f"{merge_request_path}/changes"
            )
            if payload.get("overflow"):
                payload = await self._client.request(
                    "GET",
                    f"{merge_request_path}/changes",
                    params={"access_raw_diffs": "true"},
                )
                if payload.get("overflow"):
                    raise GitLabError(
                        422,
                        "GitLab returned a truncated merge request diff",
                        "diff_truncated",
                    )
            raw_files = payload.get("changes", [])
        else:
            if any(not file.get("diff") for file in raw_files):
                raw_files = (
                    await self._raw_changes(merge_request_path) or raw_files
                )
        return [DiffFile.model_validate(file) for file in raw_files]

File: web/src/gitlab/diffs.ts
import type { DiffFile, MergeRequest } from '../api/types'
import type { GitLabClient } from './client'
import { GitLabError } from './errors'

export function mergeRequestPath(project: string, iid: number): string {
  return `/projects/${encodeURIComponent(project)}/merge_requests/${iid}`
}

/** GitLab omits these booleans; the Pydantic models default them to false. */
function normaliseFile(raw: Record<string, unknown>): DiffFile {
  return {
    old_path: String(raw.old_path ?? ''),
    new_path: String(raw.new_path ?? ''),
    diff: String(raw.diff ?? ''),
    new_file: Boolean(raw.new_file),
    renamed_file: Boolean(raw.renamed_file),
    deleted_file: Boolean(raw.deleted_file),
    collapsed: Boolean(raw.collapsed),
    too_large: Boolean(raw.too_large),
  }
}

export async function getMergeRequest(
  client: GitLabClient,
  project: string,
  iid: number,
  signal?: AbortSignal,
): Promise<MergeRequest> {
  return client.request<MergeRequest>('GET', mergeRequestPath(project, iid), {
    signal,
  })
}

/**
 * Read every diff straight from the repository.
 *
 * GitLab applies its diff size budget across the whole merge request, so files
 * past the budget come back without any content and without a `too_large` or
 * `collapsed` flag. `access_raw_diffs` bypasses the budget. Returns null when
 * the retry is unavailable, leaving the truncated diffs in place rather than
 * losing them.
 */
async function rawChanges(
  client: GitLabClient,
  path: string,
  signal?: AbortSignal,
): Promise<Record<string, unknown>[] | null> {
  try {
    const payload = await client.request<{
      overflow?: boolean
      changes?: Record<string, unknown>[]
    }>('GET', `${path}/changes`, {
      params: { access_raw_diffs: 'true' },
      signal,
    })
    if (payload.overflow) return null
    return payload.changes?.length ? payload.changes : null
  } catch {
    return null
  }
}

export async function getDiffs(
  client: GitLabClient,
  project: string,
  iid: number,
  signal?: AbortSignal,
): Promise<DiffFile[]> {
  const path = mergeRequestPath(project, iid)
  let raw: Record<string, unknown>[]
  try {
    raw = await client.paginate<Record<string, unknown>>(`${path}/diffs`, signal)
    if (raw.some((file) => !file.diff)) {
      raw = (await rawChanges(client, path, signal)) ?? raw
    }
  } catch (error) {
    if (
      !(error instanceof GitLabError) ||
      (error.status !== 404 && error.status !== 500)
    ) {
      throw error
    }
    let payload = await client.request<{
      overflow?: boolean
      changes?: Record<string, unknown>[]
    }>('GET', `${path}/changes`, { signal })
    if (payload.overflow) {
      payload = await client.request('GET', `${path}/changes`, {
        params: { access_raw_diffs: 'true' },
        signal,
      })
      if (payload.overflow) {
        throw new GitLabError(
          422,
          'GitLab returned a truncated merge request diff',
          'diff_truncated',
        )
      }
    }
    raw = payload.changes ?? []
  }
  return raw.map(normaliseFile)
}
